File: modules/ballerina-core/src/main/java/org/wso2/ballerina/core/parser/visitor/VariableDeclarationVisitor.java
/*
 * Copyright (c) 2016, WSO2 Inc. (http://wso2.com) All Rights Reserved.
 * <p>
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.wso2.ballerina.core.parser.visitor;

import org.wso2.ballerina.core.interpreter.SymbolTable;
import org.wso2.ballerina.core.model.SymbolName;
import org.wso2.ballerina.core.model.VariableDcl;
import org.wso2.ballerina.core.model.types.Type;
import org.wso2.ballerina.core.parser.BallerinaBaseVisitor;
import org.wso2.ballerina.core.parser.BallerinaParser;

/**
 * Visitor for variable declaration
 */
public class VariableDeclarationVisitor extends BallerinaBaseVisitor {
    private SymbolTable variableDeclarationSymbolTable;

    public VariableDeclarationVisitor(SymbolTable parentSymbolTable) {
        this.variableDeclarationSymbolTable = new SymbolTable(parentSymbolTable);
    }
    /**
     * {@inheritDoc}
     * <p>
     * <p>The default implementation returns the result of calling
     * {@link #visitChildren} on {@code ctx}.</p>
     *
     * @param ctx
     */
    @Override
    public Object visitVariableDeclaration(BallerinaParser.VariableDeclarationContext ctx) {
        TypeNameVisitor typeNameVisitor = new TypeNameVisitor();
        return new VariableDcl((Type) ctx.typeName().accept(typeNameVisitor),
                new SymbolName(ctx.Identifier().getText()), null);
    }

    /**
     * Base method for retrieving the symbol table
     *
     * @return symbol table for this instance
     */
    @Override
    public SymbolTable getSymbolTable() {
        return this.variableDeclarationSymbolTable;
    }
}

File: modules/web/js/ballerina/visitors/source-gen/variable-declaration-visitor.js
/**
 * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
define(['require','lodash', 'log', 'event_channel', './abstract-source-gen-visitor', '../../ast/module'],
    function(require, _, log, EventChannel, AbstractSourceGenVisitor, AST) {

        var VariableDeclarationVisitor = function(parent){
            AbstractSourceGenVisitor.call(this,parent);
        };

        VariableDeclarationVisitor.prototype = Object.create(AbstractSourceGenVisitor.prototype);
        VariableDeclarationVisitor.prototype.constructor = VariableDeclarationVisitor;

        VariableDeclarationVisitor.prototype.canVisitVariableDeclaration = function (variableDeclaration) {
            return true;
        };

        VariableDeclarationVisitor.prototype.beginVisitVariableDeclaration = function (variableDeclaration) {
            this.appendSource(variableDeclaration.getType() + " " +variableDeclaration.getIdentifier());
            log.debug('Begin Visit Variable Declaration');
        };

        VariableDeclarationVisitor.prototype.visitVariableDeclaration = function (variableDeclaration) {
            log.debug('Visit Variable Declaration');
        };

        VariableDeclarationVisitor.prototype.endVisitVariableDeclaration = function (variableDeclaration) {
            this.appendSource(";\n");
            this.getParent().appendSource(this.getGeneratedSource());
            log.debug('End Visit Variable Declaration');
        };
        
        return VariableDeclarationVisitor;
    });